### PR TITLE
Sandbox implementation

### PR DESCRIPTION
**Pull request by marcus-ares**
This PR introduces a sandbox implementation.

The purpose of the sandbox is to reduce the impact of a Firedancer compromise.

When starting, before executing any task code and/or processing user-provided input, a Firedancer process prepares everything it needs in order to function properly. This initialization mostly consists in:

- [Creating new tiles](https://github.com/firedancer-io/firedancer/blob/fc95073e1de66796732e679fad3a7d29ff0003b9/src/util/tile/fd_tile_threads.cxx#L479) which will be used for task execution (clone syscall)
- Opening the relevant workspaces which will be used to perform work and communicate (mmap syscall)

Immediately after performing those operations, Firedancer sandboxes itself. Note that firedancer has to be started as *root* in order to be able to sandbox itself. Here are the mechanisms currently used by Firedancer to achieve sandboxing:

- clearenv
    - **What:** The environment variable are cleared.
    - **Why:** Environment variables are commonly used to hold secrets. If Firedancer is compromised, no secrets living in the operator’s environment will be leaked.
    - Note: Future desired behavior: `FD_` prefixed environment variables should not be cleared
- fd_jail_setup_netns
    - **What (simplified):** The process looses access to network interfaces.
    - **What:** The process joins a previously configured network namespace. Most processes (exceptions are metrics & RPC) join an empty namespace (loopback is present).
    - **Why:** Principle of least privilege: in the event where the process was able to interact with a network interface, it should not be able to perform any communication.
- fd_jail_setup_mountns
    - **What (simplified):** The process gets a restricted view of the filesystem.
    - **What:** The ends up in a mountns with a root of its own. There is a bind mount at /log for storing logs. (log mount is not yet impl.)
    - **Why:** Principle of least privilege: the process should only be able to interact with files that it needs to function. Although we could do away with the filesystem altogether (passing log file handles through uds), we believe that it’s okay to grant component access to a mount for the purpose of logging.
- fd_jail_set_resource_limits
    - **What:** For the time being, the number of concurrently opened files is set to a low number. In the future, more resources types can be limited.
    - **Why:** Firedancer processes have well understood expected behaviors and resource needs. A process should not be able to exceed those limits, potentially leading to availability issues.
- fd_jail_set_user
    - **What:** The process drops all of its {real and effective} {user and group} ids to those of a predefined user (or those of nobody).
    - **Why:** In the case where another control was to fail, the process should be interacting with the system as an unprivileged user.
- fd_jail_close_non_std_fds
    - **What:** All file descriptors above a specified number are forcefully closed.
    - **Why:** Similar to and more impactful than `clearenv`, an operator’s process can have FDs opened that are 1. not relevant to Firedancer 2. references to sensitive resources. Those resources should not be made available to Firedancer.
- fd_jail_seccomp:
    - **What:** Prevent the usage of most syscall only allowing those explicitly needed by the specific Firedancer component.
    - **Why:** Syscalls are used to interact with the operating system. [There exists close to 400 syscalls](https://github.com/torvalds/linux/blob/v4.17/arch/x86/entry/syscalls/syscall_64.tbl). While running, A Firedancer process requires [14 syscalls](https://github.com/firedancer-io/firedancer/blob/696158aecaefa43d689e585667af534bcf76007b/src/util/sandbox/fd_sandbox.c#L212-L233) out of those 400 in order to perform its functions. Firedancer will crash if it attempts to use a syscall that is not expected.
    - Note: It also happens that the syscalls that Firedancer is using are ubiquitous and well understood. They have stood the test of time (not that time is an ultimate metric for greatness). We have the luxury of disallowing all of the syscalls that might have received less scrutiny.

### CODE REVIEW

**Review by ripatel-fd — COMMENTED** [review]
This is awesome! Would it be possible to split into multiple PRs perhaps? (e.g. starting with seccomp, then doing the namespaces next) I only had the chance to give it a superficial look so far, but added a few comments. Will review more in depth tomorrow.

File: src/util/sandbox/test_sandbox.c
Comment (marcus-jump) [reply src/util/sandbox/test_sandbox.c:355]: I'll fix the function signature.

I will call fd_boot and fd_halt in the test children if that's ok with you.

**Comment by marcus-ares** [comment]
https://github.com/firedancer-io/firedancer/pull/252#pullrequestreview-1396932066

I'd rather not split it unless you think it provides value or strongly feel that it's how it should happen. I'd prefer spending my time elsewhere if possible.

File: src/app/frank/fd_frank_init
Comment (marcus-jump) [reply src/app/frank/fd_frank_init:88]: I initially thought that we'd need users to gate access to different workspaces but it turns out we don't need users for that. Since shmem is access through hugetblfs, the process can mmap the needed pages and then sandbox itself.
It won't be able to:
1- use mmap (seccomp)
2- see the different wksps since they aren't mounted in its new mountns

File: src/util/sandbox/test_sandbox.c
Comment (SEJeff) [inline src/util/sandbox/test_sandbox.c:217]: Is this `exit(0)` superfluous due to the `exit(0)` on line 220?

File: src/util/sandbox/test_sandbox.c
Comment (SEJeff) [inline src/util/sandbox/test_sandbox.c:228]: Similar comments for the other tests would be nice!

File: src/util/sandbox/test_sandbox.c
Comment (SEJeff) [inline src/util/sandbox/test_sandbox.c:287]: If I understand this test correctly, it is just checking the truthiness of the `st_uid` and `st_gid` of the statbuf. Does it make sense to be explicit and check for `sb.st_uid == 0` and `sb.st_gid == 0`?

**Review by SEJeff — DISMISSED** [review]
Really awesome work! Just some minor comments (and I meant to hit Comment not Approve).

File: src/util/sandbox/test_sandbox.c
Comment (marcus-jump) [reply src/util/sandbox/test_sandbox.c:287]: https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdbool.h.html

It's just checking if it's 0 as 0 is the only "false" value. IIRC, there are a lot of other places in FD where comparisons like this are made so I just consider it idiomatic. I'll defer to @ripatel-jump to know if the correction is needed.

Comment (SEJeff) [reply src/util/sandbox/test_sandbox.c:287]: That works. I just like the python approach of "explicit is better than implicit", though I'll always defer to whatever is idiomatic for the language and project. Likely no correction at all.

File: src/util/sandbox/test_sandbox.c
Comment (SEJeff) [inline src/util/sandbox/test_sandbox.c:157]: nasty

I love it :)

File: src/app/frank/fd_frank_mon.bin.c
Comment (marcus-jump) [reply src/app/frank/fd_frank_mon.bin.c:366]: Bad string replace

**Review by SEJeff — DISMISSED** [review]
Please make sure to get a review from someone else on the team, but this seems sensible to me and I've confirmed the tests build and run successfully!

File: src/util/sandbox/fd_sandbox.h
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox.h:27]: Getting further along here, feels like there wants to be a sandbox profile that just exists by default for each application thread group (e.g. Linux PID).  On targets without FD_HAS_SANDBOX, this profile would be initialized during fd_boot (after the log has gone live) to indicate to the thread group there is no sandboxing.

Similarly, on targets with FD_HAS_SANDBOX, this profile would be set to sane defaults if no parameters were specified (e.g. no sandboxing) and admin defaults parameters passed by the command line (stripped) / environment. otherwise.

Then then would be some additional APIs that allow sandboxing aware code to further restrict their permissions after they do initialization.  (E.g. setup all shared memory pages and then drop various syscall permissions so that running code can't change any internal memory region firewalling.)

(The boot and and app requested changes should be extensively logged as well to aid in debugging sandboxing, understanding fine grained permissions, etc.)

File: src/util/sandbox/fd_sandbox_private.h
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox_private.h:1]: Ditto

File: src/util/sandbox/fd_sandbox_private.h
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox_private.h:6]: A private internal header makes more sense here for an include like this.  But this is also redundant with the previous header.  And generally private headers are used for all sorts of implementation specific details that should not be public but are needed to coordinate across multiple compile units.  So, there is a good chance this doesn't need to be here either as per the previous discussion.

File: src/util/sandbox/fd_sandbox_private.h
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox_private.h:14]: Not clear if these APIs should be private given the above.  E.g. could have the defaults set during boot.  And then if an application ways to do additional restrictions, let them (and just have these APIs fail and log attempts to request less restrictive permissions than they currently have).

See note previously about "struct sock_fprog".

fd_sandbox_change_user might have implications on fd_log_user functionality there.  Need some thought there what this implies.

File: src/util/sandbox/test_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/test_sandbox.c:3]: See notes above about #ifndef, fd_util_base.c, etc.

File: src/util/sandbox/test_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/test_sandbox.c:5]: Likewise, this is probably unnecessary given the tweaks above and perhap usage of FD_HAS_SANDBOX in the Local.mk (such that the test is only built on targets where it is necessary).

File: src/util/sandbox/test_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/test_sandbox.c:11]: Aspirationally, only the top level include (e.g. fd_util.h) would be necessary for a unit tester and and the APIs to test would be such that all functionality is testable public facing APIs.

Given all the low level OS hacker, this is probably not possible here.  But, we probably can get very close and at a minimum here, we shouldn't have to include fd_sandbox.h and fd_sandbox_private.h  (e.g. would it ever make sense to just include fd_sandbox_private.h standalone ... doubt it ... generally private APIs imply the public API will be available too).

File: src/util/sandbox/fd_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox.c:8]: See notes about private implies public.

File: src/util/sandbox/fd_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox.c:24]: Willing to bet there is a bunch of cleanups and compression that can be done here.  (Also, the linux/filter.h is redundant with previous, but this is actually the place where it actually should be.)

File: src/util/sandbox/fd_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox.c:26]: Should be done by public facing API as per previous comments.

File: src/util/sandbox/fd_sandbox.c
Comment (kbowers-jump) [inline src/util/sandbox/fd_sandbox.c:85]: See notes about fd_log_user.  No objection here but how fd_log_user behaves in conjunction with this API should be sorted.

File: src/util/sandbox/fd_sandbox.c
Comment (kbowers-jump) [reply src/util/sandbox/fd_sandbox.c:34]: Also, if sandbox is util, the netns name probably should be be frank app specific.  This is probably something that comes from the command line / environment on fd_boot (with maybe an option to set it programmatically post boot if it wasn't already set on fd_boot).

File: src/app/frank/fd_frank_main.c
Comment (marcus-jump) [reply src/app/frank/fd_frank_main.c:128]: I don't think that we want to automatically sandbox everything from the start. The initialization might require more permissions as it will do things such as mapping other workspaces or setting up xdp. We could probably have two sandboxing steps (pre-init, post-init), but as long as we can consider init "safe", we can just live with post-init sandboxing.

File: config/with-sandbox.mk
Comment (mmcgee-jump) [reply config/with-sandbox.mk:1]: Really don't like sandboxing being optional or having "HAS_SANDBOX". Everything should be sandboxed 100% of the time, it's part of the application.

**Review by mmcgee-jump — COMMENTED** [review]
Took an initial pass -- some comments inline

**Review by mmcgee-jump — CHANGES_REQUESTED** [review]
Took an initial pass -- some comments inline

File: src/util/sandbox/fd_sandbox.c
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox.c:60]: I was on the fence about doing it but yes, it costs us nothing.

File: src/app/frank/fd_frank_init
Comment (marcus-jump) [reply src/app/frank/fd_frank_init:87]: The logs that end up on the host system need to be owned by a sensible user. That can all go away if / when we change how logs are being emitted. We can cross that bridge once we get there.

File: config/with-sandbox.mk
Comment (marcus-jump) [reply config/with-sandbox.mk:1]: Kevin and I also sync on it last week. Some platforms won't support sandboxing. That said, there should be a generic sandbox implementation for unsupported platforms that just makes you pass a flag that makes the operator acknowledge that they are doing something bad.

File: src/util/sandbox/fd_sandbox.c
Comment (mmcgee-jump) [reply src/util/sandbox/fd_sandbox.c]: I realized this will be a small problem at the moment since this is how QUIC variables are passed at the moment.

**Comment by marcus-ares** [comment]
https://app.codecov.io/github/firedancer-io/firedancer/commit/a7c6c86839e56e375ef1b320a7191fe73b3cd13b

Looks like coverage of forked processes is not computed.

File: src/util/sandbox/fd_sandbox.c
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox.c]: artifact of debugging :)

**Comment by mmcgee-jump** [comment]
Can you also add a `security.md` to the docs folder describing the security architecture? Could just be basically the same thing as this change overview.

**Comment by marcus-ares** [comment]
@mmcgee-jump -- This is planned for an upcoming PR. Thanks for mentioning it!

File: src/util/sandbox/fd_sandbox_linux_private.h
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox_linux_private.h:4]: I feel like this is just going to get out of sync, better in a `security-architecture.md` file and then point to that from here.

File: src/util/sandbox/fd_sandbox.h
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox.h:33]: Don't link to github here, just reference the code itself.

File: src/app/frank/fd_frank_main.c
Comment (mmcgee-jump) [inline src/app/frank/fd_frank_main.c:40]: nit: extra whitespace

File: src/util/sandbox/fd_sandbox_linux.c
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox_linux.c:137]: Not sure why the network namespace needs to be named and set up / destoyed by the init script? I think you can just unshare a new NETNS here and have it be anonymous?

File: src/app/frank/fd_frank_init
Comment (mmcgee-jump) [inline src/app/frank/fd_frank_init:84]: See comment where this gets used

File: src/util/sandbox/fd_sandbox_linux.c
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox_linux.c:37]: This also should be an anonymous user, not named or set up before hand. See chatgpt response for a question like "Can I create an anonymous user  in C code that isn't named, for use with a user namespace?"

File: src/util/sandbox/fd_sandbox.h
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox.h:8]: This is apparently not used anywhere?

File: src/util/sandbox/fd_sandbox_linux.c
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox_linux.c:157]: Don't you also need to remove all effective caps from the current process as well?

File: src/util/sandbox/fd_sandbox_linux.c
Comment (mmcgee-jump) [inline src/util/sandbox/fd_sandbox_linux.c:163]: It doesn't appear this argument is ever non-null? Can we just remove it?

**Review by mmcgee-jump — CHANGES_REQUESTED** [review]
Few more comments inline

File: src/util/test_util_base.c
Comment (barnjamin) [inline src/util/test_util_base.c:20]: duplicate?

File: src/util/sandbox/fd_sandbox_linux_private.h
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox_linux_private.h:4]: I can remove it once we have security-architecture.md.

File: src/util/sandbox/fd_sandbox_linux.c
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox_linux.c:135]: You could decide to have it live in a preconfigured netns. It's probably what the RPC task is going to require.

File: src/util/sandbox/fd_sandbox.h
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox.h:8]: Oh yes, true! We need to talk about. We might need to have a special mode for the quic tile when the driver operates in skb mode. The current reference is inaccurate and I'll just remove it for now.

File: src/util/sandbox/fd_sandbox_linux.c
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox_linux.c:155]: I'm revisiting this whole thing. Thanks for pointing it out! I think that we need to sort out the tile zero issue out first before we can handle caps during init properly.

File: src/util/sandbox/fd_sandbox_linux.c
Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox_linux.c:161]: It is not right now but it was made for making it easy for users to pick where the chroot lives. In hindsight, this is an operator concern.

I have greater plans for how we handle mountns in the future so I'm not against removing this argument.

Comment (marcus-jump) [reply src/util/sandbox/fd_sandbox_linux.c]: Argument removed :)

**Pull request closed without merging**
